service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AgentDirectoryMapper.java

```diff
@@ -0,0 +1,141 @@
+package uk.nhs.adaptors.gp2gp.ehr.mapper;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.StringUtils;
+import org.hl7.fhir.dstu3.model.Bundle;
+import org.hl7.fhir.dstu3.model.Observation;
+import org.hl7.fhir.dstu3.model.Organization;
+import org.hl7.fhir.dstu3.model.Patient;
+import org.hl7.fhir.dstu3.model.Practitioner;
+import org.hl7.fhir.dstu3.model.PractitionerRole;
+import org.hl7.fhir.dstu3.model.ReferralRequest;
+import org.hl7.fhir.dstu3.model.ResourceType;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import com.github.mustachejava.Mustache;
+
+import lombok.RequiredArgsConstructor;
+import uk.nhs.adaptors.gp2gp.ehr.exception.EhrMapperException;
+import uk.nhs.adaptors.gp2gp.ehr.mapper.parameters.AgentDirectoryParameter;
+import uk.nhs.adaptors.gp2gp.ehr.utils.ResourceExtractor;
+import uk.nhs.adaptors.gp2gp.ehr.utils.TemplateUtils;
+
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+@Component
+public class AgentDirectoryMapper {
+    private static final Mustache AGENT_DIRECTORY_STRUCTURE_TEMPLATE = TemplateUtils.loadTemplate("ehr_agent_structure_template.mustache");
+    private static final Map<String, Boolean> MAPPED_ORGANIZATIONS_AND_PRACTITIONER = new HashMap<>();
+
+    private final PractitionerAgentPersonMapper practitionerAgentPersonMapper;
+    private final OrganizationToAgentMapper organizationToAgentMapper;
+
+    public String mapEHRFolderToAgentDirectory(Bundle bundle, String nhsNumber) {
+        var builder = AgentDirectoryParameter.builder()
+            .patientManagingOrganization(extractPatientManagingOrganization(bundle, nhsNumber))
+            .observationOrganizations(prepareObservationAgentsList(bundle))
+            .referralRequestsOrganizations(prepareReferralRequestAgentsList(bundle))
+            .agentPersons(preparePractitionerRoleAgentsList(bundle));
+
+        MAPPED_ORGANIZATIONS_AND_PRACTITIONER.clear();
+        return TemplateUtils.fillTemplate(AGENT_DIRECTORY_STRUCTURE_TEMPLATE, builder.build());
+    }
+
+    private String extractPatientManagingOrganization(Bundle bundle, String nhsNumber) {
+        Optional<Patient> patientOptional = AgentDirectoryExtractor.extractPatientByNhsNumber(bundle, nhsNumber);
+        if (patientOptional.isPresent() && patientOptional.get().hasManagingOrganization()) {
+            var organization = ResourceExtractor.extractResourceByReference(bundle,
+                patientOptional.get().getManagingOrganization().getReferenceElement());
+            if (organization.isPresent()) {
+                return organizationToAgentMapper.mapOrganizationToAgent((Organization) organization.get());
+            }
+        }
+        throw new EhrMapperException("No patient or managing organization found in ehrFolder with NHS Number: " + nhsNumber);
```

Comment:
This could actually all become functional
```suggestion
    private String extractPatientManagingOrganization(Bundle bundle, String nhsNumber) {
        return AgentDirectoryExtractor.extractPatientByNhsNumber(bundle, nhsNumber)
            .filter(Patient::hasManagingOrganization)
            .map(patient -> ResourceExtractor.extractResourceByReference(bundle,
                patient.getManagingOrganization().getReferenceElement()))
            .map(Organization.class::cast)
            .map(organizationToAgentMapper::mapOrganizationToAgent)
            .orElseThrow(() -> new EhrMapperException(
                "No patient or managing organization found in ehrFolder with NHS Number: " + nhsNumber));
    }
```